src/contracts/core/PaymentCoordinator.sol

```diff
@@ -0,0 +1,474 @@
+     */
+    function _verifyTokenClaimProof(
+        bytes32 earnerTokenRoot,
+        uint32 tokenLeafIndex,
+        bytes calldata tokenProof,
+        TokenTreeMerkleLeaf calldata tokenLeaf
+    ) internal pure {
+        // Verify inclusion of token leaf
+        bytes32 tokenLeafHash = calculateTokenLeafHash(tokenLeaf);
+        require(
+            Merkle.verifyInclusionKeccak({
+                root: earnerTokenRoot,
+                index: tokenLeafIndex,
+                proof: tokenProof,
+                leaf: tokenLeafHash
+            }),
+            "PaymentCoordinator._verifyTokenClaim: invalid token claim proof"
+        );
+    }
+
+    /**
+     * @notice verify inclusion of earner claim proof in the distribution root. This verifies
+     * the inclusion of the earner and earnerTokenRoot hash in the tree. The token claims are proven separately
+     * against the earnerTokenRoot hash (see _verifyTokenClaimProof). The earner leaf comprises of (earner, earnerTokenRoot)
+     * @param root distribution root that should be read from storage
+     * @param earnerLeafIndex index of the earner leaf
+     * @param earnerProof proof of the earners account root in the merkle tree
+     * @param earnerLeaf leaf of earner merkle tree containing the earner address and earner's token root hash
+     */
+    function _verifyEarnerClaimProof(
+        bytes32 root,
+        uint32 earnerLeafIndex,
+        bytes calldata earnerProof,
+        EarnerTreeMerkleLeaf calldata earnerLeaf
+    ) internal pure {
+        // Verify inclusion of earner leaf
+        bytes32 earnerLeafHash = calculateEarnerLeafHash(earnerLeaf);
+        require(
+            Merkle.verifyInclusionKeccak({
+                root: root,
+                index: earnerLeafIndex,
+                proof: earnerProof,
+                leaf: earnerLeafHash
+            }),
+            "PaymentCoordinator._verifyEarnerClaimProof: invalid earner claim proof"
+        );
+    }
+
+    /**
+     * @notice Sets the address of the entity that can claim payments on behalf of the earner
+     * @param claimer The address of the entity that can claim payments on behalf of the earner
+     * @dev Only callable by the `earner`
+     */
+    function setClaimerFor(address claimer) external {
+        address earner = msg.sender;
+        address prevClaimer = claimerFor[earner];
+        claimerFor[earner] = claimer;
+        emit ClaimerForSet(earner, prevClaimer, claimer);
+    }
+
+    /// @notice return the hash of the earner's leaf
+    function calculateEarnerLeafHash(EarnerTreeMerkleLeaf calldata leaf) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(leaf.earner, leaf.earnerTokenRoot));
+    }
+
+    /// @notice returns the hash of the earner's token leaf
+    function calculateTokenLeafHash(TokenTreeMerkleLeaf calldata leaf) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(leaf.token, leaf.cumulativeEarnings));
+    }
```

Comment:
We should consider adding some specific delimiters to each of these hash calculations.

i.e:

```
uint8 EARNER_LEAF = 1;
uint8 TOKEN_LEAF = 2;

earnerLeafHash = keccak(EARNER_LEAF, earner, tokenRoot);

tokenLeafHash = keccak(TOKEN_LEAF, token, cumulativeEarnings);
```

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/514/commits/24fb70726a39a125e61cc19a361d7cc20aa7b930
Added and updated backend claimgen as well